espup: init at 0.4.1

knightpp opened the pull request:
## Description of changes

This PR adds https://github.com/esp-rs/espup tool. It's meant to download rust ESP toolchain into home directory. Note, it only works when rustup is installed.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

knightpp committed:
espup: init at 0.4.1

knightpp committed:
maintainers: add knightpp

drupol commented on pkgs/development/tools/espup/default.nix:
```diff
@@ -0,0 +1,60 @@
+
+  cargoHash = "sha256-GYhF6VDBAieZbu4x9EiQVVJkmx0aRYK0xwGGP0nuVGc=";
+
+  preBuild = ''
+    export OPENSSL_NO_VENDOR="1"
+  '';
```
Move this part after `nativeBuildInputs` and `buildInputs`.

drupol commented on pkgs/development/tools/espup/default.nix:
```diff
@@ -0,0 +1,60 @@
+  preCheck = ''
+    export HOME=$(mktemp -d)
+  '';
+  checkFlags = [
```
Keep it consistent with the rest of the file, add an empty line here.
```suggestion

  checkFlags = [
```

drupol commented on pkgs/top-level/all-packages.nix:
```diff
@@ -41544,4 +41544,5 @@ with pkgs;
 
   wttrbar = callPackage ../applications/misc/wttrbar { };
 
+  espup = callPackage ../development/tools/espup { };
```
This doesn't seem to be at the right place, please sort it out.

knightpp replied:
Sorry, I didn't see categories prefixed with `###`. Fixed.

drupol reviewed (changes requested):
Hi !

Thanks for your contribution!

I've left some comments, left me know if this is clear enough.

drupol commented on pkgs/development/tools/espup/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+  src = fetchFromGitHub {
+    owner = "esp-rs";
+    repo = pname;
```
`repo = pname;` is nice for DRY but creates a binding that goes too far.

See further information about this here: https://github.com/nix-community/nixpkgs-lint/issues/21

knightpp commented:
Hi. I've addressed your comments, thanks! Is `rec` at the top level of package considered bad? I can replace it with `let ... in` but not sure what's more appropriate.

drupol commented:
`rec` is not considered bad, but we can usually replace it with a better construction.

Sadly, this is not available when using the `rustPlatform.buildRustPackage` wrapper.

I would leave it as it is, it's seems to be ok, I'm doing local tests and do my feedback in a couple of minutes.

drupol reviewed (changes requested):
Hi !

Thanks for your first contribution to the Nix project!

I used a tool to generate the nix file for the `espup` project and there are differences, could you please check it out and do the necessary changes?

Here's the file:

<details>

```nix
{ lib
, rustPlatform
, fetchFromGitHub
, pkg-config
, bzip2
, openssl
, xz
, zstd
, stdenv
, darwin
}:

rustPlatform.buildRustPackage rec {
  pname = "espup";
  version = "0.4.1";

  src = fetchFromGitHub {
    owner = "esp-rs";
    repo = "espup";
    rev = "v${version}";
    hash = "sha256-gzM+RT4Rt+LaYk7CwYUTIMci8DDI0y3+7y+N2yKRDOc=";
  };

  cargoHash = "sha256-GYhF6VDBAieZbu4x9EiQVVJkmx0aRYK0xwGGP0nuVGc=";

  nativeBuildInputs = [
    pkg-config
  ];

  buildInputs = [
    bzip2
    openssl
    xz
    zstd
  ] ++ lib.optionals stdenv.isDarwin [
    darwin.apple_sdk.frameworks.Security
  ];

  env = {
    OPENSSL_NO_VENDOR = true;
    ZSTD_SYS_USE_PKG_CONFIG = true;
  };

  meta = with lib; {
    description = "Tool for installing and maintaining Espressif Rust ecosystem";
    homepage = "https://github.com/esp-rs/espup/";
    license = with licenses; [ asl20 mit ];
    maintainers = with maintainers; [ ];
  };
}
```

</details>

To generate the file on your computer, run: `nix run nixpkgs#nix-init -- -u https://github.com/esp-rs/espup/`

knightpp commented:
Fixed according to the `nix-init` auto-generated file. Also, added shell completions.

drupol reviewed (approved):
top !
